crates/accelerate/src/linear_matrix.rs

```diff
@@ -0,0 +1,111 @@
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use ndarray::{s, ArrayViewMut2, Axis};
+use numpy::PyReadwriteArray2;
+use pyo3::prelude::*;
+use rayon::prelude::*;
+
+// Perform ROW operation on a matrix mat
+fn _row_op(mat: &mut ArrayViewMut2<bool>, ctrl: usize, trgt: usize) {
+    let row0 = mat.row(ctrl).to_owned();
+    let mut row1 = mat.row_mut(trgt);
+    row1.zip_mut_with(&row0, |x, &y| *x ^= y);
+}
+
+// Perform COL operation on a matrix mat
+fn _col_op(mat: &mut ArrayViewMut2<bool>, ctrl: usize, trgt: usize) {
+    let col0 = mat.column(ctrl).to_owned();
+    let mut col1 = mat.column_mut(trgt);
+    col1.zip_mut_with(&col0, |x, &y| *x ^= y);
+}
+
+// Gauss elimination of a matrix mat with m rows and n columns.
+// If full_elim = True, it allows full elimination of mat[:, 0 : ncols]
+// Returns the matrix mat.
+fn gauss_elimination(mut mat: ArrayViewMut2<bool>, ncols: Option<usize>, full_elim: Option<bool>) {
+    let (m, mut n) = (mat.nrows(), mat.ncols()); // no. of rows and columns
+    if let Some(ncols_val) = ncols {
+        n = usize::min(n, ncols_val); // no. of active columns
+    }
+
+    let mut r = 0; // current rank
+    let k = 0; // current pivot column
+    let mut new_k = 0;
+    while (r < m) && (k < n) {
+        let mut is_non_zero = false;
+        let mut new_r = r;
+        for j in k..n {
+            new_k = k;
+            for i in r..m {
+                if mat[(i, j)] {
+                    is_non_zero = true;
+                    new_k = j;
+                    new_r = i;
+                    break;
+                }
+            }
+            if is_non_zero {
+                break;
+            }
+        }
+        if !is_non_zero {
+            return; // A is in the canonical form
+        }
+
+        if new_r != r {
+            let temp_r = mat.slice_mut(s![r, ..]).to_owned();
+            let temp_new_r = mat.slice_mut(s![new_r, ..]).to_owned();
+            mat.slice_mut(s![r, ..]).assign(&temp_new_r);
+            mat.slice_mut(s![new_r, ..]).assign(&temp_r);
+        }
+
+        // Copy source row to avoid trying multiple borrows at once
+        let row0 = mat.row(r).to_owned();
+        if full_elim == Some(true) {
+            mat.axis_iter_mut(Axis(0))
+                .into_par_iter()
+                .enumerate()
+                .filter(|(i, row)| (*i < r) && row[new_k])
+                .for_each(|(_i, mut row)| {
+                    row.zip_mut_with(&row0, |x, &y| *x ^= y);
+                });
+        }
+
+        mat.axis_iter_mut(Axis(0))
+            .into_par_iter()
+            .enumerate()
+            .filter(|(i, row)| (*i > r) && (*i < m) && row[new_k])
+            .for_each(|(_i, mut row)| {
+                row.zip_mut_with(&row0, |x, &y| *x ^= y);
+            });
```

Comment:
This can be simplified to a single iterator with something like (double check that I got the filter condition correct:

```suggestion
        if {
            mat.axis_iter_mut(Axis(0))
                .into_par_iter()
                .enumerate()
                .filter(|(i, row)| (full_elim == Some(true)  && (*i < r) && row[new_k]) || (*i > r && *i < m))
                .for_each(|(_i, mut row)| {
                    row.zip_mut_with(&row0, |x, &y| *x ^= y);
                });
        }
```

You can also move the `full_elim` condition into the `for_each` block and just change the filter to be `i < m` if that's easier.

Reply:
The other thing I was thinking about is you can probably get a slice view of the rows you care about and iterate over that slice instead of the full matrix. It's might be worth investigating that instead of doing a single big iterator like this. TBH, I'm not sure what is going to be the most efficient here we need to benchmark all the different possible ways to write this and figure out what ends up being the best performing.